GitCommands/Patch/Patch.cs

```diff
@@ -1,56 +1,63 @@
-using System.Text;
+﻿using System;
 using JetBrains.Annotations;
 
 namespace PatchApply
 {
-    public class Patch
+    public sealed class Patch
     {
-        private StringBuilder _textBuilder;
+        [NotNull]
+        public string Header { get; }
 
-        public Patch()
-        {
-            File = FileType.Text;
-        }
+        [CanBeNull]
+        public string Index { get; }
 
-        public enum PatchType
-        {
-            NewFile,
-            DeleteFile,
-            ChangeFile,
-            ChangeFileMode
-        }
+        public PatchFileType FileType { get; }
 
-        public enum FileType
-        {
-            Binary,
-            Text
-        }
+        [NotNull]
+        public string FileNameA { get; }
 
-        public string PatchHeader { get; set; }
-        public string PatchIndex { get; set; }
-        public FileType File { get; set; }
-        public string FileNameA { get; set; }
-        public string FileNameB { get; set; }
-        public bool IsCombinedDiff { get; set; }
+        [CanBeNull]
+        public string FileNameB { get; }
 
-        public PatchType Type { get; set; }
+        public bool IsCombinedDiff { get; }
 
-        [CanBeNull]
-        public string Text => _textBuilder?.ToString();
+        public PatchChangeType ChangeType { get; }
 
-        public void AppendText(string text)
+        [CanBeNull]
+        public string Text { get; }
+
+        public Patch(
+            [NotNull] string header,
+            [CanBeNull] string index,
+            PatchFileType fileType,
+            [NotNull] string fileNameA,
+            [CanBeNull] string fileNameB,
+            bool isCombinedDiff,
+            PatchChangeType changeType,
+            [CanBeNull] string text)
         {
-            GetTextBuilder().Append(text);
+            Header = header ?? throw new ArgumentNullException(nameof(header));
+            Index = index;
+            FileType = fileType;
+            FileNameA = fileNameA ?? throw new ArgumentNullException(nameof(fileNameA));
+            FileNameB = fileNameB;
+            IsCombinedDiff = isCombinedDiff;
+            ChangeType = changeType;
+            Text = text;
         }
+    }
 
-        public void AppendTextLine(string line)
-        {
-            GetTextBuilder().Append(line).Append('\n');
-        }
+    public enum PatchChangeType
```

Comment:
:bulb: Top-level type should be in its own file

Reply:
There are many violations of this. These types were already comingling. If this is something we want as policy in the project, let's enforce it via SCA.

I'm not a huge fan of enforcing it though. Rarely do I navigate to a file through the file system view, and I don't think I'm alone. Having related files next to one another is useful in some cases. In this case the whole file with all three types is 63 lines long. The enums explain what state a `Patch` can be in, and I like having all of that visible at a glance. Just my £0.02.

As an aside, I worked on a C++ code base where every single _method_ was defined in its own source file, and every class in its own folder, with subclasses in subfolders. That was a nightmare! Maybe I'm just traumatised from that experience.

Reply:
I just like how it works with GitHub (Press <kbd>T</kbd> to navigate to file) and the Filter files view in GitExtensions. It means I always know where to find things. :smile: